pkg/controller/populators/populator-base.go

```diff
@@ -0,0 +1,335 @@
+		if !errors.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, nil
+	}
+	return pvcPrime, nil
+}
+
+func (r *ReconcilerBase) getPV(volName string) (*corev1.PersistentVolume, error) {
+	pv := &corev1.PersistentVolume{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: volName}, pv); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return pv, nil
+}
+
+func (r *ReconcilerBase) handleStorageClass(pvc *corev1.PersistentVolumeClaim) (bool, bool, error) {
+	if pvc.Spec.StorageClassName == nil {
+		return false, false, nil
+	}
+
+	waitForFirstConsumer := false
+	storageClass, err := cc.GetStorageClassByName(r.client, pvc.Spec.StorageClassName)
+	if err != nil {
+		return false, waitForFirstConsumer, err
+	}
+
+	if checkIntreeStorageClass(pvc, storageClass) {
+		r.log.V(2).Info("can't use populator for PVC with in-tree storage class", "namespace", pvc.Namespace, "name", pvc.Name, "provisioner", storageClass.Provisioner)
+		return false, waitForFirstConsumer, nil
+	}
+
+	if storageClass.VolumeBindingMode != nil && *storageClass.VolumeBindingMode == storagev1.VolumeBindingWaitForFirstConsumer {
+		waitForFirstConsumer = true
+		nodeName := pvc.Annotations[cc.AnnSelectedNode]
+		if nodeName == "" {
+			// Wait for the PVC to get a node name before continuing
+			return false, waitForFirstConsumer, nil
+		}
+	}
+
+	return true, waitForFirstConsumer, nil
+}
+
+func (r *ReconcilerBase) createPVCPrime(pvc *corev1.PersistentVolumeClaim, source client.Object, waitForFirstConsumer bool, pvcModifier pvcModifierFunc) (*corev1.PersistentVolumeClaim, error) {
+	labels := map[string]string{
+		common.CDILabelKey: common.CDILabelValue,
+	}
+	annotations := make(map[string]string)
+	annotations[cc.AnnImmediateBinding] = ""
+	if waitForFirstConsumer {
+		annotations[cc.AnnSelectedNode] = pvc.Annotations[cc.AnnSelectedNode]
+	}
+	pvcPrimeName := PVCPrimeName(pvc)
+	pvcPrime := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        pvcPrimeName,
+			Namespace:   pvc.Namespace,
+			Labels:      labels,
+			Annotations: annotations,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes:      pvc.Spec.AccessModes,
+			Resources:        pvc.Spec.Resources,
+			StorageClassName: pvc.Spec.StorageClassName,
+			VolumeMode:       pvc.Spec.VolumeMode,
+		},
+	}
+	pvcPrime.OwnerReferences = []metav1.OwnerReference{
+		*metav1.NewControllerRef(pvc, schema.GroupVersionKind{
+			Group:   "",
+			Version: "v1",
+			Kind:    "PersistentVolumeClaim",
+		}),
+	}
+	util.SetRecommendedLabels(pvcPrime, r.installerLabels, "cdi-controller")
+
+	if !isPVCOwnedByDataVolume(pvc) {
+		requestedSize, _ := pvc.Spec.Resources.Requests[corev1.ResourceStorage]
+		// disk or image size, inflate it with overhead
+		requestedSize, err := cc.InflateSizeWithOverhead(r.client, requestedSize.Value(), &pvc.Spec)
+		if err != nil {
+			return nil, err
```

Comment:
since you moved the event to after the create fails should we add here also a log or event?

Reply:
You are right, didn't notice that having the event after the function allowed us to trigger it with more error cases. Will move it back.